crawl/crawl.go

```diff
@@ -1,178 +1,218 @@
+	endIndex := int64(sth.TreeSize - 1)
+
 	fetcher := scanner.NewFetcher(
 		logClient,
 		&scanner.FetcherOptions{
 			BatchSize:     *batchSize,
-			StartIndex:    *startIndex,
-			EndIndex:      *endIndex,
+			StartIndex:    0,
+			EndIndex:      endIndex,
 			ParallelFetch: *numWorkers,
 			Continuous:    false,
 		},
 	)
 
-	// Create a callback that sends fetched entries to a channel to be processed.
-	entries := make(chan entryData)
-	processBatch := func(batch scanner.EntryBatch) {
-		for i, e := range batch.Entries {
-			index := batch.Start + int64(i)
-			rawLogEntry, err := ct.RawLogEntryFromLeaf(index, &e)
-			if err != nil {
-				fmt.Printf("failed to process entry at index %d: %v\n", index, err)
-				continue
-			}
-			entries <- entryData{index: batch.Start + int64(i), entry: rawLogEntry}
+	// Fetch the entries as fast as possible
+	entries := runFetcher(fetcher)
+	// Filter them into sorted order
+	sortedEntries := sorter(entries)
+
+	hasher := rfc6962.DefaultHasher
+	compactRange := (&compact.RangeFactory{Hash: hasher.HashChildren}).NewEmptyRange(0)
+
+	start := time.Now()
+	var rootHash []byte
+	var e entryData
+	for e = range sortedEntries {
+		compactRange.Append(e.merkleLeafHash, nil)
+		rootHash, err = compactRange.GetRootHash(nil)
+		if err != nil {
+			// The only way this could happen is if we didn't start
+			// the range at 0, which we definitely did.
+			log.Fatalf("index %d: %s", e.index, err)
+		}
+		elapsed := time.Since(start).Seconds()
+		fraction := float64(e.index) / float64(endIndex)
+		rate := time.Duration(float64(e.index) / float64(elapsed))
+		var eta time.Duration
+		if rate > 0 {
+			eta = time.Duration(endIndex-e.index) / rate * time.Second
 		}
-	}
-
-	// Start a worker which reads entries from the fetchers. If it receives an
-	// entry for which it has not already processed the immediately preceding
-	// entry, it buffers it and waits to process it until the intervening entries
-	// have been backfilled. When processing an entry, it compares it to the
-	// immediately prior entry: if the timestamp has traveled backwards in time,
-	// it outputs the *prior* entry, on the assumption that its timestamp was
-	// incorrectly forward in time.
-	buffer := NewHeap[entryData]()
-	nextIndex := *startIndex
-	last := entryData{index: -1}
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for e := range entries {
-			// If it's not the entry we want, save it for later.
-			if e.index != nextIndex {
-				buffer.Push(e)
-				continue
-			}
-
-			// If this is the first entry we're looking for, initialize our last-seen
-			// tracker so we have a basis for comparison.
-			if last.index == -1 {
-				last = e
-				nextIndex = e.index + 1
-				continue
-			}
 
-			// This is the next entry we were looking for. Process it.
-			processEntry(last, e)
-			last = e
-			nextIndex += 1
-			if nextIndex%1000 == 0 {
-				fmt.Printf("Processed up to index %d\n", nextIndex)
-			}
+		if e.index%98689 == 0 {
```

Comment:
Where's this modulus come from?

Reply:
I figured if we're always doing multiples of, say, 100,000, we might miss some patterns of trees because we're always verifying even-sized trees. Picked a prime to be cute. :-)